### ISSUE TITLE
Snail page - order not going through

### ISSUE THREAD

**Issue by kirstenalarsen** [openfoodfoundation/openfoodnetwork#4222]
## Description

Reported issue from user of occurrences today of orders not going through and customer getting snail page. They 

one sent me this screen shot of the message that appeared when they tried to submit their order at the checkout. It was between 12pm - 1.45pm this afternoon.

They were able to order in the order cycle ( they were ordering under different oc days 2.3719 and 1.3719) but when they submitted their order the error page appeared. Their order didn’t come through- and wan’t saved. 

They tried several times. I said maybe they should refresh the screen.  Thought it might be a server issue/ one was on Microsoft Explorer (Edge?). Not sure about the 2nd person. 
The majority of our customers were fine.
I was able to send through orders with out any problem when I tested it.

NB. Enterprise ID 153. This is the same enterprise that had the problem last night with the cloning order cycles #4216 . . is there any possibility this is related?

## Expected Behavior
Order should go through

## Actual Behavior
It didn't

## Steps to Reproduce
We don't know how to replicate. Hoping logs will shed light

## Severity
S1 - placing orders is pretty critical! Only 2 known customers, but who knows if others not reporting it

**Comment by kirstenalarsen** [comment]
@mkllnk looked at logs and identified 2 known errors (2 orders affected). One of them looked like a 'missing billing first name' and the other something about paypal. I have: 
* confirmed that the user does not have a paypal payment method
* confirmed that if i leave the billing first name empty I can't place an order, and get correctly pointed up to the empty billing name field as you'd expect

http://recordit.co/kyLX9NQvIb

**Comment by kirstenalarsen** [comment]
also for context, while in Wormly seeing if this could be related to UK problems in #4210 we discovered that Aus memory is not ok - - see wormly

**Comment by luisramos0** [comment]
in bugsnag I can see 14 errors related to paypal express payments not working in the last 2 days. They are all from different users but all are for the same payment method: 273 which is a paypal express method with description "Bank transfer(...)" from enterprise 2173 Wit_chie Har_vests

The other error "Cannot transition state via :restart_checkout from :payment" is known but this time is coming from checkout#edit and not checkout#update as usual.
The discovery is that these errors coming from checkout#edit are from the same users (email is the same) as the paypal express errors above.

Conclusion, the checkout errors in bugsnag are payment errors related to paypal express method 273.
Is this issue related to this enterprise?
If yes, there's a problem with that payment method.
If no, these bugsnag errors are not related to this issue.

**Comment by kirstenalarsen** [comment]
Thanks @luisramos0 - I'm pretty sure that's a different user - looking into it. But are you telling me this happened 14 times yesterday rather than the 2 we know about? And any hints on what is wrong W the payment method?

**Comment by kirstenalarsen** [comment]
OK, so Witchie Harvest is a completely different enterprise from the one that reported the issue, so I think the errors you are looking at are not the ones connected to the reported issue. I have however checked this out and they had a payment method set to paypal accidentally I think, and I've confirmed that I get the snail when I try to use it - it had no paypal credentials but did include bank transfer details. I have changed it to cash/eft method

Can you tell me the names / emails of the users affected? as I assume they will have got the snail - and we should contact them and apologise and ask them to try and order again.

In the meantime, still don't know about the food connect ones

**Comment by kirstenalarsen** [comment]
ALSO I have now confirmed that even now that I have fixed the payment method I am getting snail when I try to checkout. This happens even after I have emptied the cart and filled it again AND after I have logged out and in again

http://recordit.co/tws5bj5msR

So I am not confident emailing these users yet because they might still not be able to order? Help!

**Comment by kirstenalarsen** [comment]
OK, on further investigation this is getting worse and worse. On the hunch that perhaps once this problem was triggered it was causing problems elsewhere, I have now tried ordering in another shop, and as another user in another shop and all I get is SNAIL SNAIL SNAIL 

https://recordit.co/1pQLwad70P
IT IS UNCLEAR TO ME WHETHER ANYONE CAN PLACE AN ORDER on Aus Prod at the moment. We are waking up for a busy day of orders on a Friday . . all hands on deck please?!

**Comment by kirstenalarsen** [comment]
If the times are correct in orders, the last order on OFN Aus was placed last night at 11:43pm . . this could well be correct, but I would expect to see orders coming through soon. And I currently can't place one

**Comment by kirstenalarsen** [comment]
ok so when I open a new browser and new session I can place an order - so hopefully people coming fresh to the system this morning can place orders. But even closing the window completely and reopening in firefox I CANNOT check out in the browser where I encountered the payment method problem

**Comment by kirstenalarsen** [comment]
I now have some orders through and am seeing some from other people, so we're not totally down. I am very uncomfortable 

It looks like the paypal problem triggers the state machine transition problem, which might then affect all users for a certain period of time . . 

![image](https://user-images.githubusercontent.com/2911176/64386989-fc116a80-d07d-11e9-8f05-4a27765652e3.png)

I really need to know what is happening. Can we tell people to try again or will they just have the same problem?

**Comment by luisramos0** [comment]
there was a deployment going on when you were testing @kirstenalarsen

**Comment by kirstenalarsen** [comment]
there was a deploy of 2.4.1 (yay) at around the time I was having this problem, but I’m not sure the deploy totally answers it . . because i was going through lots of steps as in logging in and out, going into different shops, creating orders etc and demonstrating snail on checkout in different situations. and the bugsnag thing i posted shows the repitition of state transition engine . . so actually on reflection i don’t think the deploy was the problem. That would have meant i couldn’t load anything

**Comment by kirstenalarsen** [comment]
It should be possible to recreate this on a staging server quite easily by just creating a paypal express payment method with no details and trying to place an order on it. I'm genuinely sad that I need to go now and can't do this!

**Comment by luisramos0** [comment]
I am not sure I follow: the paypal express were with Wit_chie Har_vests but you say this issue is with another enterprise. if this other enterprise had a paypal express like Wit_chie Har_vests we would get bugsnag alerts, which we dont, we only have alerts for Wit_chie Har_vests paypal payment method.

**Comment by emilyjeanrogers** [comment]
Is there any link between the 14 errors that @luisramos0 picked up in bugsnag and the paypal error that @mkllnk picked up in the logs relating to paypal payment method?

As far I can understand from the comments above, the bugsnag errors relate to Wit_chie Har_vests and the paypal payment method that @mkllnk mentioned relate to Food Connect (Enterprise ID 153). 

There's no link between the two enterprises, so were they both impacted by the same issue, or 
are we dealing with two different issues?

**Comment by kirstenalarsen** [comment]
As far as I can see or understand Food Connect has no paypal method, so it is very strange that they managed to trigger a paypal error. Perhaps worth checking with them @emilyjeanrogers if there is any possibility that anyone accidentally had an incorrectly set up paypal method for any period yesterday? Otherwise, it seems like somehow what was happeing with Witchie somehow affected Food Cpnnect. Which does seem to be what I replicated above

**Comment by emilyjeanrogers** [comment]
Food Connect (Enterprise ID 153) have just reported that they had another customer experience the same issue (order not going through and customer getting snail page) on Wednesday night 4th Sep, but were then able to order normally on Thursday 5th Sep.

The initial reported issue occurred on Thursday 5th Sep between 12pm - 2pm.

**Comment by mkllnk** [comment]
This issue has been linked to an error in Bugsnag
[**NoMethodError** in **spree/paypal#express**](https://app.bugsnag.com/yaycode/openfoodnetwork/errors/5d6ea59aa104a600184d9e67?event_id=5d719db8004f3daef7c90000&i=gh&m=li)

**Comment by mkllnk** [comment]
I think the Paypal failure has been introduced by c56486d7aeaee822adc93f1f6d43be282a178a89 in #3670 first released in [v2.2.0](https://github.com/openfoodfoundation/openfoodnetwork/releases/tag/v2.2.0) (July).

**Comment by mkllnk** [comment]
It also looks like that this is connected to another problem. When something in the payment doesn't work, we try to reset the order and go back to the checkout page. But resetting the order seems to be broken in a lot of cases: https://github.com/openfoodfoundation/openfoodnetwork/issues/4005

**Comment by luisramos0** [comment]
@mkllnk the paypal bugsnag error is related to the Witchie enterprise, not food connect.
as far as I can see theres no connection between those paypal express errors and this issue in Food connect...

**Comment by luisramos0** [comment]
I agree, to fix the paypal express error (the checkout would fail, it's just the recovery so that user does not see the snail if paypal payment fails, user should be redirected to checkout page) we need to re-add this route:
`get '/checkout/:state', :to => 'checkout#edit', :as => :checkout_state`

I'll create a separate issue for this paypal express error now.

**Comment by luisramos0** [comment]
in terms of #4005, that is also a problem with recovery from a failed payment, so resolving 4005 will not resolve the fact that the checkout fails but users will see the checkout page with an error instead of the snail.

So, I think here we have two scenarios where payment fails and user sees the snail instead of being redirected back to the checkout page.

**Comment by luisramos0** [comment]
I am downgrading this to s2 (s3 maybe?) it looks like this is just users having trouble checking out with their payment methods and OFN showing the snail instead of saying "failed payment": the 2 issues described in previous comments.

**Comment by mkllnk** [comment]
This issue has been unlinked from an error in Bugsnag
[**NoMethodError** in **spree/paypal#express**](https://app.bugsnag.com/yaycode/openfoodnetwork/errors/5d6ea59aa104a600184d9e67?event_id=5d71ccd4004f30f03d9e0000&i=gh&m=ul)

**Comment by luisramos0** [comment]
this unlinking above was my doing in bugsnag. I'll open a new issue and link the bugsnag error to that issue.

**Comment by mkllnk** [comment]
@luisramos0 I think as well that it's not related to Food Connect. Kirsten was quite sure that when she encountered the problem with Witchie that she then couldn't order at another enterprise with a new user and in a different browser. But in our last conversation she said that it was rushed it maybe the other browser still had access to the broken cart. I would assume that Food Connect is not affected by this unless someone can re-produce this scenario.

**Comment by mkllnk** [comment]
@kirstenalarsen I think that this issue can be closed and replaced by #4005 and #4233 which mean "snail on checkout page" and "snail when Paypal is wrong". Do you agree?

**Comment by emilyjeanrogers** [comment]
I'm adding the below details of how I replicated the issue that @kirstenalarsen was experiencing, which demonstrates that the payment error in one shop causes breakage for the same user, in another shop, even when using a different browser. This could be tested further to see if it causes breakage for a different user / different shop as well. 

On Sept 6th between 10 - 11am.  

1. Created an open OC for test Enterprise (Kirsten Test Farm) on Staging https://staging.openfoodnetwork.org.au/kristen-test-farm/shop
2. Set payment method as Paypal, without valid paypal details (@kirstenalarsen can you expand on how you created this broken payment method) 
3. Accessed store for Kirsten Test Farm, attempted to checkout
4. Triggered Snail on checkout
5. Created an order cycle for another Enterprise (Emilys Seedling Shop).
6. Tried to place an order with the second test shop, same user, same browser
7. Triggered Snail - can't process order.
8. Opened Emily's Seedling Shop in a different browser, same user and tried to place an order
9. Triggered snail on checkout.

Note: I wasn’t able to go on and test whether this would impact a different user in a different shop, as I couldn't set up a second user on staging (confirmation emails not working).

**Comment by luisramos0** [comment]
I see that the solution to #4233 is already deployed in staging AUS. I am not sure this was the case when you performed the test above @emilyjeanrogers 

I gave it a try now:

point 3. I tried to buy apples from kirsten test farm, flash error after pressing checkout:
![Screenshot 2019-09-16 at 14 37 28](https://user-images.githubusercontent.com/1640378/64962545-88e2d200-d88f-11e9-8918-b4596e7c961b.png)
This replicates the error in point 4 above but now shows a good error message ✅ 

point 5. Then I went to Emilys Seedling Shop to buy something, I got a snail on loading the checkout page itself...
This was caused by a broken stripe payment method. The stripe payment method didnt have the enterprise selected. When I go to the backoffice and edit the payment method, I try to update and get the correct error message: "Stripe account owner can't be blank". This make me conclude that this was caused by the data migration from live environment: when you create a stripe payment method the page forces you to select an enterprise, so this will not happen anywhere else.
I selected another payment method for emilys seedling shop and I managed to checkout.
This is maybe the replication of point 7 above? I am not sure are I didnt even see the checkout page, did you see the checkout page and pressed the checkout button?

Summary: the first error was due to broken paypal express payment method, no snail now after #4233 , the second error was due to a stripe payment method not well configured due to data migrated and truncate from a live database.

I think we can close this issue.

**Comment by luisramos0** [comment]
Moving to test ready. Feel free to move to in dev if you think there's something else to be fixed or if you think we should investigate further.

**Comment by daniellemoorhead** [comment]
@kirstenalarsen have assigned you to this alongside of @emilyjeanrogers as the two of you are best placed to retest and know via @luisramos0 comment if it needs more work.

**Comment by daniellemoorhead** [comment]
From @emilyjeanrogers:

>I've asked Kirsten if she's happy to close it, as she raised the initial issue. I don't know if it's resolved as I don't fully understand the solution - and it would be very time consuming to set up the re-testing scenario. My suggestion would be to close it, and only re-visit if it comes up again?

So @luisramos0 let's consider this done for now, and it can be reopened if needs be later on. I'll move it to ready to go, not sure if it can be closed or not.

**Comment by luisramos0** [comment]
ok, thanks. yeah, we can close, it's an issue.
